tests/FieldTest.php

```diff
@@ -889,4 +889,77 @@ public function testDateTimeFieldsToString()
         $this->assertEquals($dt->format('H:i:s.u'), $model->getField('time')->toString());
         $this->assertEquals($dt->format('Y-m-d\TH:i:s.uP'), $model->getField('datetime')->toString());
     }
+
+    public function testSetNull()
+    {
+        $m = new Model();
+        $m->addField('a');
+        $m->addField('b', ['mandatory' => true]);
+        $m->addField('c', ['required' => true]);
+
+        // valid value for set()
+        $m->set('a', 'x');
+        $m->set('b', 'y');
+        $m->set('c', 'z');
+        $this->assertSame('x', $m->get('a'));
+        $this->assertSame('y', $m->get('b'));
+        $this->assertSame('z', $m->get('c'));
+        $m->set('a', '');
+        $m->set('b', '');
+        $this->assertSame('', $m->get('a'));
+        $this->assertSame('', $m->get('b'));
+        $m->set('a', null);
+        $this->assertNull($m->get('a'));
+
+        // null must pass
+        $m->setNull('a');
+        $m->setNull('b');
+        $m->getField('c')->setNull();
+        $this->assertNull($m->get('a'));
+        $this->assertNull($m->get('b'));
+        $this->assertNull($m->get('c'));
+
+        // invalid value for set() - normalization must fail
+        $this->expectException(\atk4\data\Exception::class);
+        $m->set('c', null); // @TODO even "b"/mandatory field should fail!
```

Comment:
@DarkSide666 Am I right, that with "b" field / mandatory field this should fail too?

Reply:
Yes `b` (`mandatory`) should also not allow `null` value, but allows empty string value.
Basically `mandatory` should be called `notNull` and `required` should be called `notEmpty`, but changing these property names would be big compatibility break and a lot of code to refactor again in actual projects which use atk4/data :(
Also one of these properties (not sure now which one it was) is more like UI specific property. I guess it's `required` (notEmpty) one, but again, changing that will break a lot of projects.

Reply:
@DarkSide666 Yes, I always need to look at the phpdoc for this. `notNull` and `notEmpty` will be perfect - ok to refactor? This project is still in very early stage. If we want to make this project big, I think we should not wait with these changes.

To the issue - can you propose a PR with fix before this is merged? The `mandatory` normalization/validation currently does not work.

Reply:
Yes `notNull` and `notEmpty` would be perfect names for these properties, but changing that will break almost every single data model we have in all our projects :(
Can we change them in at least semi-compatible manner? Maybe introduce `isNotNull()` and `isNotEmpty()` methods for that and also use both property names just to keep some compatibility.
On the other hand of course it would be much better (cleaner code) to just change these property names everywhere. Also would like to have @romaninsh input on this. Sadly he's quite hard to catch lately.

Reply:
@DarkSide666 The current design choose is actually a slightly different - `required` does not allow everything that evalueated with `empty()` to true, thus even not null, I understand that it may be handy to set one `required` property, but it is horrible design choice as `mandatory` can be false even if the field is effectively `mandatory`.

Rejecting empty values but still allowing null may be perfectly valid, at least in `atk4/data`.

Wdyt and what is your prefference? Two bool properties `notNull` and `notEmpty` ("empty" excl. null) or one enum property for all 4 possibilities?

Reply:
I do not see 4 possibilities here - only two.
* only restrict `null` values - everything else (empty string, zero, false etc.) is allowed
* restrict `null` *and* empty values - do not allow empty string, false, (not sure about this - i think zero should be allowed - it's legit value) zero, empty array etc.

Also there was idea that actually one of these properties more belong to UI than Data level, but it would be quite painful to move it somewhere else because of backward compatibility...

Reply:
@DarkSide666 - no restriction, prevent null, prevent empty but allow null, prevent null and empty

Reply:
_prevent empty but allow null_ - I can't imagine of any use-case for this.

Reply:
@DarkSide666 In data it is very common - null has meaning of N/A, not set yet, while empty means empty value is desired - and not always empty value should be allowed.